File: Chapter 4 (Local search)/Basic/Python/first_choice_hc.py
import random

def first_choice_hc(f,neighbors,init,max_steps):
    x=init()
    fx=f(x)
    for _ in range(max_steps):
        nbr=list(neighbors(x))
        random.shuffle(nbr)
        improved=False
        for y in nbr:
            fy=f(y)
            if fy<fx:
                x=y
                fx=fy
                improved=True
                break
        if not improved:
            break
    return x,fx

if __name__=="__main__":
    def f(x): return (x-3)*(x-3)
    def neighbors(x): return [x-1,x+1]
    init=lambda: random.randint(-10,10)
    print(first_choice_hc(f,neighbors,init,100))

File: Chapter 4 (Local search)/Basic/C++/first_choice_hc.cpp
#include <bits/stdc++.h>
using namespace std;
int f(int x) { return (x - 3) * (x - 3); }
vector<int> neighbors(int x) { return {x - 1, x + 1}; }
pair<int, int> first_choice_hc(int (*F)(int), vector<int> (*N)(int), int init, int max_steps)
{
    mt19937 rng(123);
    int x = init;
    int fx = F(x);
    for (int step = 0; step < max_steps; step++)
    {
        auto nbr = N(x);
        shuffle(nbr.begin(), nbr.end(), rng);
        bool improved = false;
        for (int y : nbr)
        {
            int fy = F(y);
            if (fy < fx)
            {
                x = y;
                fx = fy;
                improved = true;
                break;
            }
        }
        if (!improved)
            break;
    }
    return {x, fx};
}
int main()
{
    auto res = first_choice_hc(f, neighbors, 0, 100);
    cout << res.first << " " << res.second << "\n";
}
